File: test/bench/test_controlled_multiplication_modulo_n.cpp
/*
 * Copyright (c) 2023 - 2026 Chair for Design Automation, TUM
 * Copyright (c) 2025 - 2026 Munich Quantum Software Company GmbH
 * All rights reserved.
 *
 * SPDX-License-Identifier: MIT
 *
 * Licensed under the MIT License
 */

#include "bench/ControlledMultiplicationModuloN.hpp"
#include "bench/Evaluation.hpp"

#include <gtest/gtest.h>

#include <array>
#include <cstddef>
#include <stdexcept>
#include <string>
#include <string_view>

namespace {

using mqt::bench::ControlledMultiplicationModuloN;
using mqt::bench::ControlledMultiplicationModuloNOptions;
using mqt::bench::Counts;
using mqt::bench::Output;

TEST(ControlledMultiplicationModuloN, StoresParametersAndOutput) {
  const ControlledMultiplicationModuloN benchmark{
      {.multiplier = "011", .modulus = "101"}};
  EXPECT_EQ(benchmark.options().multiplier, "011");
  EXPECT_EQ(benchmark.options().modulus, "101");
  EXPECT_EQ(benchmark.output(), (Output{"result", 8}));
}

TEST(ControlledMultiplicationModuloN, ValidatesTheConfiguredInstance) {
  const auto maximumMultiplier =
      std::string(ControlledMultiplicationModuloNOptions::MAX_BITS - 1U, '0') +
      "1";
  const auto maximumModulus =
      "1" +
      std::string(ControlledMultiplicationModuloNOptions::MAX_BITS - 1U, '0');
  EXPECT_NO_THROW(static_cast<void>(ControlledMultiplicationModuloN{
      {.multiplier = maximumMultiplier, .modulus = maximumModulus}}));

  EXPECT_THROW(static_cast<void>(ControlledMultiplicationModuloN{
                   {.multiplier = "0", .modulus = "1"}}),
               std::invalid_argument);
  EXPECT_THROW(static_cast<void>(ControlledMultiplicationModuloN{
                   {.multiplier = "001", .modulus = "1000"}}),
               std::invalid_argument);
  EXPECT_THROW(static_cast<void>(ControlledMultiplicationModuloN{
                   {.multiplier = "00x", .modulus = "101"}}),
               std::invalid_argument);
  EXPECT_THROW(static_cast<void>(ControlledMultiplicationModuloN{
                   {.multiplier = "001", .modulus = "10x"}}),
               std::invalid_argument);
  EXPECT_THROW(static_cast<void>(ControlledMultiplicationModuloN{
                   {.multiplier = "001", .modulus = "011"}}),
               std::invalid_argument);
  EXPECT_THROW(static_cast<void>(ControlledMultiplicationModuloN{
                   {.multiplier = "000", .modulus = "101"}}),
               std::invalid_argument);
  EXPECT_THROW(static_cast<void>(ControlledMultiplicationModuloN{
                   {.multiplier = "101", .modulus = "101"}}),
               std::invalid_argument);
  EXPECT_THROW(static_cast<void>(ControlledMultiplicationModuloN{
                   {.multiplier = "110", .modulus = "101"}}),
               std::invalid_argument);

  const auto tooLongMultiplier =
      std::string(ControlledMultiplicationModuloNOptions::MAX_BITS, '0') + "1";
  const auto tooLongModulus =
      "1" + std::string(ControlledMultiplicationModuloNOptions::MAX_BITS, '0');
  EXPECT_THROW(
      static_cast<void>(ControlledMultiplicationModuloN{
          {.multiplier = tooLongMultiplier, .modulus = tooLongModulus}}),
      std::invalid_argument);
}

TEST(ControlledMultiplicationModuloN,
     GivesUniformWeightToTheExactControlledProducts) {
  const ControlledMultiplicationModuloN benchmark{
      {.multiplier = "011", .modulus = "101"}};
  constexpr std::array<std::string_view, 8> multiplicands{
      "000", "001", "010", "011", "100", "101", "110", "111",
  };
  constexpr std::array<std::string_view, 8> products{
      "000", "011", "001", "100", "010", "000", "011", "001",
  };

  Counts exact;
  for (size_t index = 0; index < multiplicands.size(); ++index) {
    const auto inactive =
        std::string{"0"} + std::string{multiplicands[index]} + "0000";
    const auto active = std::string{"1"} + std::string{multiplicands[index]} +
                        "0" + std::string{products[index]};
    EXPECT_DOUBLE_EQ(benchmark.probability(inactive), 1. / 16.);
    EXPECT_DOUBLE_EQ(benchmark.probability(active), 1. / 16.);
    exact.emplace(inactive, 1U);
    exact.emplace(active, 1U);
  }

  EXPECT_DOUBLE_EQ(benchmark.probability("10010000"), 0.);
  EXPECT_DOUBLE_EQ(benchmark.probability("01110001"), 0.);
  EXPECT_THROW(static_cast<void>(benchmark.probability("1001001")),
               std::invalid_argument);
  EXPECT_THROW(static_cast<void>(benchmark.probability("100100x1")),
               std::invalid_argument);

  const auto evaluation = benchmark.evaluate(exact);
  EXPECT_DOUBLE_EQ(evaluation.totalVariationDistance, 0.);
  EXPECT_DOUBLE_EQ(evaluation.squaredHellingerFidelity, 1.);
  EXPECT_FALSE(evaluation.successProbability);
}

TEST(ControlledMultiplicationModuloN,
     SupportsNonCoprimeInputsAndMultiplicandsAtLeastN) {
  const ControlledMultiplicationModuloN nonCoprime{
      {.multiplier = "010", .modulus = "100"}};
  EXPECT_DOUBLE_EQ(nonCoprime.probability("10110010"), 1. / 16.);

  const ControlledMultiplicationModuloN benchmark{
      {.multiplier = "011", .modulus = "101"}};
  EXPECT_DOUBLE_EQ(benchmark.probability("11110001"), 1. / 16.);
}

TEST(ControlledMultiplicationModuloN,
     KeepsTheLargestReferenceWeightRepresentable) {
  constexpr auto width = ControlledMultiplicationModuloNOptions::MAX_BITS;
  const auto multiplier = std::string(width - 1U, '0') + "1";
  const auto modulus = "1" + std::string(width - 1U, '0');
  const ControlledMultiplicationModuloN benchmark{
      {.multiplier = multiplier, .modulus = modulus}};
  const auto outcome =
      "1" + std::string(width, '1') + "00" + std::string(width - 1U, '1');
  EXPECT_GT(benchmark.probability(outcome), 0.);
}

} // namespace

File: test/python/bench/test_controlled_multiplication_modulo_n.py
# Copyright (c) 2023 - 2026 Chair for Design Automation, TUM
# Copyright (c) 2025 - 2026 Munich Quantum Software Company GmbH
# All rights reserved.
#
# SPDX-License-Identifier: MIT
#
# Licensed under the MIT License

"""Python bindings for the controlled multiplication modulo N benchmark."""

from __future__ import annotations

import json

import pytest

from mqt.core.bench import controlled_multiplication_modulo_n

from .utils import assert_generates


def _make_benchmark() -> controlled_multiplication_modulo_n.ControlledMultiplicationModuloN:
    return controlled_multiplication_modulo_n.ControlledMultiplicationModuloN(
        controlled_multiplication_modulo_n.Options(multiplier="011", modulus="101")
    )


def _exact_counts() -> dict[str, int]:
    products = ("000", "011", "001", "100", "010", "000", "011", "001")
    counts: dict[str, int] = {}
    for multiplicand, product in zip((f"{value:03b}" for value in range(8)), products, strict=True):
        counts[f"0{multiplicand}0000"] = 1
        counts[f"1{multiplicand}0{product}"] = 1
    return counts


def test_controlled_multiplication_modulo_n_reference() -> None:
    """Expose the options, output, and exact modular products."""
    benchmark = _make_benchmark()
    assert benchmark.options.multiplier == "011"
    assert benchmark.options.modulus == "101"
    assert benchmark.output.name == "result"
    assert benchmark.output.width == 8
    for outcome in _exact_counts():
        assert benchmark.probability(outcome) == pytest.approx(1 / 16)
    assert benchmark.probability("10010000") == 0


def test_controlled_multiplication_modulo_n_evaluation() -> None:
    """Evaluate exact counts against the modular-product reference."""
    evaluation = _make_benchmark().evaluate(_exact_counts())
    assert evaluation.total_variation_distance == pytest.approx(0)
    assert evaluation.squared_hellinger_fidelity == pytest.approx(1)
    assert evaluation.success_probability is None


def test_controlled_multiplication_modulo_n_json_roundtrip() -> None:
    """Preserve the benchmark identity through both JSON representations."""
    benchmark = _make_benchmark()
    assert json.loads(benchmark.instance_specification_json)["parameters"] == {
        "modulus": "101",
        "multiplier": "011",
    }

    instance_copy = controlled_multiplication_modulo_n.ControlledMultiplicationModuloN.from_instance_specification_json(
        benchmark.instance_specification_json
    )
    manifest_copy = controlled_multiplication_modulo_n.ControlledMultiplicationModuloN.from_manifest_json(
        benchmark.manifest_json
    )
    assert instance_copy.case_id == manifest_copy.case_id == benchmark.case_id


def test_controlled_multiplication_modulo_n_generation() -> None:
    """Generate a controlled modular-multiplication program."""
    assert_generates(_make_benchmark().generate())
